File: csharp/Core/CommandHandler.cs
namespace Aurelion.Connector.Core;

public sealed class CommandHandler
{
    private readonly string _instanceId;
    private readonly IOperationExecutor _ops;
    private readonly ConnectorLogger _logger;
    private readonly Action<Dictionary<string, object?>, Dictionary<string, object?>> _responseSender;

    public CommandHandler(string instanceId, IOperationExecutor ops,
                          ConnectorLogger logger,
                          Action<Dictionary<string, object?>, Dictionary<string, object?>> responseSender)
    {
        _instanceId = instanceId;
        _ops = ops;
        _logger = logger;
        _responseSender = responseSender;
    }

    public void Handle(Dictionary<string, object?> command)
    {
        var correlationId = AsString(command, "correlation_id");
        var operation = AsString(command, "operation");
        var resultStorageRequested = command.TryGetValue("result_storage_requested", out var rsr)
                                    && (rsr is true || (rsr is string s && s.Equals("true", StringComparison.OrdinalIgnoreCase)));
        var payload = command.TryGetValue("payload", out var p) && p is Dictionary<string, object?> dict
            ? dict : new Dictionary<string, object?>();

        if (correlationId is null)
        {
            _responseSender(command, ErrorResponse(null, "correlation_id not provided"));
            return;
        }
        if (string.IsNullOrWhiteSpace(operation))
        {
            _responseSender(command, ErrorResponse(correlationId, "operation is required"));
            return;
        }

        var trace = ParseCommandTrace(command);
        string initT, initI, tgtT, tgtI;
        string? causationParent = null;
        if (trace is not null)
        {
            initT = trace.Value.InitiatorType;
            initI = trace.Value.InitiatorId;
            tgtT = trace.Value.TargetType;
            tgtI = trace.Value.TargetId;
            causationParent = trace.Value.ParentEventId;
        }
        else
        {
            initT = "system";
            initI = "platform";
            tgtT = "system";
            tgtI = operation;
        }

        var receivedId = _logger.Emit("info", "connector.command.received", "Command received",
            new() { ["instance_id"] = _instanceId, ["operation"] = operation },
            correlationId,
            causationId: causationParent,
            initiatorType: initT, initiatorId: initI,
            actorType: "connector", actorId: _instanceId,
            targetType: tgtT, targetId: tgtI);

        try
        {
            var result = _ops.Execute(operation, payload, resultStorageRequested, correlationId);
            var response = SuccessResponse(correlationId, result.Payload, result.StorageRef);

            _logger.Emit("info", "connector.command.completed", "Command completed",
                new() { ["instance_id"] = _instanceId, ["operation"] = operation,
                         ["stored"] = result.StorageRef is not null },
                correlationId,
                causationId: receivedId,
                initiatorType: initT, initiatorId: initI,
                actorType: "connector", actorId: _instanceId,
                targetType: tgtT, targetId: tgtI);

            _responseSender(command, response);
        }
        catch (Exception ex)
        {
            _logger.Emit("error", "connector.command.failed", "Command failed",
                new() { ["instance_id"] = _instanceId, ["operation"] = operation, ["error"] = ex.Message },
                correlationId,
                causationId: receivedId,
                initiatorType: initT, initiatorId: initI,
                actorType: "connector", actorId: _instanceId,
                targetType: tgtT, targetId: tgtI);
            _responseSender(command, ErrorResponse(correlationId, ex.Message));
        }
    }

    private record struct TraceContext(string InitiatorType, string InitiatorId,
                                         string TargetType, string TargetId, string ParentEventId);

    private static TraceContext? ParseCommandTrace(Dictionary<string, object?> command)
    {
        var raw = AsString(command, "trace_parent_event_id");
        if (string.IsNullOrWhiteSpace(raw)) return null;
        if (!Guid.TryParse(raw.Trim(), out _)) return null;

        var it = AsString(command, "trace_initiator_type");
        var ii = AsString(command, "trace_initiator_id");
        var tt = AsString(command, "trace_target_type");
        var ti = AsString(command, "trace_target_id");
        if (string.IsNullOrWhiteSpace(it) || string.IsNullOrWhiteSpace(ii)
            || string.IsNullOrWhiteSpace(tt) || string.IsNullOrWhiteSpace(ti))
            return null;

        return new TraceContext(
            it.Trim().ToLowerInvariant(),
            ii.Trim(),
            tt.Trim().ToLowerInvariant(),
            ti.Trim(),
            raw.Trim());
    }

    private static Dictionary<string, object?> ErrorResponse(string? correlationId, string message) => new()
    {
        ["correlation_id"] = correlationId,
        ["status"] = "error",
        ["payload"] = null,
        ["result_storage_ref"] = null,
        ["error"] = new Dictionary<string, object?> { ["message"] = message },
    };

    private static Dictionary<string, object?> SuccessResponse(string correlationId, object? payload,
                                                                 Dictionary<string, string>? storageRef) => new()
    {
        ["correlation_id"] = correlationId,
        ["status"] = "ok",
        ["payload"] = payload,
        ["result_storage_ref"] = storageRef,
        ["error"] = null,
    };

    private static string? AsString(Dictionary<string, object?> map, string key)
        => map.TryGetValue(key, out var val) && val is string s ? s : null;
}

File: java/src/main/java/com/aurelion/connector/core/CommandHandler.java
package com.aurelion.connector.core;

import com.aurelion.connector.core.OperationExecutor.OperationResult;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;

public final class CommandHandler {

    private final String instanceId;
    private final OperationExecutor ops;
    private final ConnectorLogger logger;
    private final MqResponseSender responseSender;

    @FunctionalInterface
    public interface MqResponseSender {
        void send(Map<String, Object> command, Map<String, Object> response);
    }

    public CommandHandler(String instanceId, OperationExecutor ops,
                          ConnectorLogger logger, MqResponseSender responseSender) {
        this.instanceId = instanceId;
        this.ops = ops;
        this.logger = logger;
        this.responseSender = responseSender;
    }

    @SuppressWarnings("unchecked")
    public void handle(Map<String, Object> command) {
        String correlationId = asString(command, "correlation_id");
        String operation = asString(command, "operation");
        boolean resultStorageRequested = Boolean.TRUE.equals(command.get("result_storage_requested"));
        Object rawPayload = command.getOrDefault("payload", Map.of());
        Map<String, Object> payload = rawPayload instanceof Map<?, ?>
                ? (Map<String, Object>) rawPayload : Map.of();

        if (correlationId == null) {
            responseSender.send(command, errorResponse(null, "correlation_id not provided"));
            return;
        }
        if (operation == null || operation.isBlank()) {
            responseSender.send(command, errorResponse(correlationId, "operation is required"));
            return;
        }

        TraceContext trace = parseCommandTrace(command);
        String initT, initI, tgtT, tgtI;
        String causationParent = null;
        if (trace != null) {
            initT = trace.initiatorType;
            initI = trace.initiatorId;
            tgtT = trace.targetType;
            tgtI = trace.targetId;
            causationParent = trace.parentEventId;
        } else {
            initT = "system";
            initI = "platform";
            tgtT = "system";
            tgtI = operation;
        }

        String receivedId = logger.emit("info", "connector.command.received", "Command received",
                Map.of("instance_id", instanceId, "operation", operation),
                correlationId, causationParent,
                initT, initI, "connector", instanceId, tgtT, tgtI);

        try {
            OperationResult result = ops.execute(operation, payload, resultStorageRequested, correlationId);
            Map<String, Object> response = successResponse(correlationId, result.payload(), result.storageRef());

            logger.emit("info", "connector.command.completed", "Command completed",
                    Map.of("instance_id", instanceId, "operation", operation,
                            "stored", result.storageRef() != null),
                    correlationId, receivedId,
                    initT, initI, "connector", instanceId, tgtT, tgtI);

            responseSender.send(command, response);
        } catch (Exception e) {
            logger.emit("error", "connector.command.failed", "Command failed",
                    Map.of("instance_id", instanceId, "operation", operation,
                            "error", e.getMessage() == null ? "" : e.getMessage()),
                    correlationId, receivedId,
                    initT, initI, "connector", instanceId, tgtT, tgtI);
            responseSender.send(command, errorResponse(correlationId, e.getMessage()));
        }
    }

    private record TraceContext(String initiatorType, String initiatorId,
                                 String targetType, String targetId,
                                 String parentEventId) {}

    private static TraceContext parseCommandTrace(Map<String, Object> command) {
        String raw = asString(command, "trace_parent_event_id");
        if (raw == null || raw.isBlank()) return null;
        try {
            UUID.fromString(raw.trim());
        } catch (IllegalArgumentException e) {
            return null;
        }
        String it = asString(command, "trace_initiator_type");
        String ii = asString(command, "trace_initiator_id");
        String tt = asString(command, "trace_target_type");
        String ti = asString(command, "trace_target_id");
        if (it == null || it.isBlank() || ii == null || ii.isBlank()
                || tt == null || tt.isBlank() || ti == null || ti.isBlank()) return null;

        return new TraceContext(
                it.trim().toLowerCase(),
                ii.trim(),
                tt.trim().toLowerCase(),
                ti.trim(),
                raw.trim()
        );
    }

    private static Map<String, Object> errorResponse(String correlationId, String message) {
        Map<String, Object> resp = new LinkedHashMap<>();
        resp.put("correlation_id", correlationId);
        resp.put("status", "error");
        resp.put("payload", null);
        resp.put("result_storage_ref", null);
        resp.put("error", Map.of("message", message == null ? "" : message));
        return resp;
    }

    private static Map<String, Object> successResponse(String correlationId,
                                                        Object payload,
                                                        Map<String, String> storageRef) {
        Map<String, Object> resp = new LinkedHashMap<>();
        resp.put("correlation_id", correlationId);
        resp.put("status", "ok");
        resp.put("payload", payload);
        resp.put("result_storage_ref", storageRef);
        resp.put("error", null);
        return resp;
    }

    private static String asString(Map<String, Object> map, String key) {
        Object val = map.get(key);
        return val instanceof String s ? s : null;
    }
}
